### ISSUE TITLE
SQL Errors in Indexing Products and Categories

### ISSUE THREAD

**Issue by richdynamix** [algolia/algoliasearch-magento-2#79]
After installing the module and configuring the admin section I have reindexed the data but I get the following error - 

```
Algolia Search Products index has been rebuilt successfully in 00:00:12
Algolia Search Categories indexer process unknown error:
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'backend.entity_id' in 'field list', query was: SELECT CONCAT(backend.store_id, '-', backend.entity_id) AS `key`, `category`.`path`, `backend`.`value` FROM `catalog_category_entity_int` AS `backend`
 INNER JOIN `catalog_category_entity` AS `category` ON backend.entity_id = category.entity_id AND (category.created_in <= 1 AND category.updated_in > 1) WHERE (backend.attribute_id = '36') ORDER BY `backend`.`store_id` ASC, `backend`.`entity_id` ASC

Algolia Search Pages index has been rebuilt successfully in 00:00:06
Algolia Search Suggestions index has been rebuilt successfully in 00:00:01
Algolia Search Suggestions index has been rebuilt successfully in 00:00:00
Algolia Search Queue Runner index has been rebuilt successfully in 00:00:00
```

I then proceeded to configure the product searchable attributes and reindex and I now get the following - 

```
Algolia Search Products indexer process unknown error:
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'backend.entity_id' in 'field list', query was: SELECT CONCAT(backend.store_id, '-', backend.entity_id), `backend`.`value` FROM `catalog_category_entity_varchar` AS `backend`
 INNER JOIN `catalog_category_entity` AS `category` ON backend.entity_id = category.entity_id AND (category.created_in <= 1 AND category.updated_in > 1) WHERE (backend.attribute_id = '35') AND (category.level > 1)
Algolia Search Categories indexer process unknown error:
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'backend.entity_id' in 'field list', query was: SELECT CONCAT(backend.store_id, '-', backend.entity_id) AS `key`, `category`.`path`, `backend`.`value` FROM `catalog_category_entity_int` AS `backend`
 INNER JOIN `catalog_category_entity` AS `category` ON backend.entity_id = category.entity_id AND (category.created_in <= 1 AND category.updated_in > 1) WHERE (backend.attribute_id = '36') ORDER BY `backend`.`store_id` ASC, `backend`.`entity_id` ASC

Algolia Search Pages index has been rebuilt successfully in 00:00:03
Algolia Search Suggestions index has been rebuilt successfully in 00:00:03
Algolia Search Suggestions index has been rebuilt successfully in 00:00:00
Algolia Search Queue Runner index has been rebuilt successfully in 00:00:00
```

Investigating the query it references `backend.entity_id` for both entities `catalog_category_entity_int` and `catalog_category_entity_varchar` however there is no `entity_id` in these tables. Updating the queries to use the corresponding `row_id` seems to return the correct data.

Issue was found in Magento 2.1.1 Enterprise Edition. It's worth noting that the data is from a partial data migration from M1 -> M2 while still in development however the table schemas have not changed.

**Comment by JanPetr** [comment]
Thank you for the report @richdynamix, will investigate and come back to you.

**Comment by richdynamix** [comment]
After further investigation last night I stumbled upon this - https://github.com/magento/magento2/issues/6478 

It seems the changing of column names is related to the content staging module in Magento 2.1 EE however I am not entirely sure how this works yet as its a new feature.

**Comment by richdynamix** [comment]
@JanPetr Is there any update on this? My client is coming to the end of their free trial with Algolia and due to this issue we have been unable to test it. We are hoping to stay with Algolia as the search service but this issues seems like a blocker at the moment.

**Comment by Xiris42** [comment]
I did fix this issue by changing the 5 occurences of 'backend.entity_id' by 'backend.row_id'. (2 in getCategoryName() and 3 in getCategories()).

File : <path_to_magento_dir>/vendor/algolia/algoliasearch-magento-2/Helper/Entity/BaseHelper.php

It seems to work fine, indexing algolia_products and algolia_categories won't trigger SQLSTATE error anymore.

**Comment by richdynamix** [comment]
@Xiris42 Do you mean you fixed this file directly? This would need to be fixed in the package otherwise it can't be updated or installed elsewhere. I would assume a check on the Magento version would be needed for this.

**Comment by Xiris42** [comment]
@richdynamix I just fixed it that way to make it work. I am just testing the extension in a local environment. Don't worry, I don't usually do horrible things like this ;)

**Comment by richdynamix** [comment]
😂 To be fair, if the issue can't be fixed in the package we might need to fork it and do a fix ourselves.

Its a bit disappointing that the content staging module would break backwards compatibility.

**Comment by JanPetr** [comment]
Hello guys,
Im sorry for being late on it. I'm right now really short on time and probably will get to this issue next week.
Thank you for your discussion and that you came up with the solution for you, that's great!

As you said, the biggest issue is the compatibility between 2.0 and 2.1. I'll try to sort it out with version detection and using the right columns. When I have a PR submitted, I'll let you know :)

Thanks!

**Comment by richdynamix** [comment]
@JanPetr If it helps I have done this - https://github.com/richdynamix/algoliasearch-magento-2/pull/1/files in a fork. While it works with my version I can't tell if it would break Magento 2.1 on CE as the content staging is suppose to be an Enterprise only feature.

I suppose knowing in that method if its CE or EE will help.

FYI Magento 2.1.2 will be released today so not sure if these bugs will be resolved, probably not but will be worth checking out.

**Comment by JanPetr** [comment]
Thanks a lot for additional information. Hopefully should get to it by the end of the week or early next week.

**Comment by JanPetr** [comment]
Hello @richdynamix and @Xiris42 for your help.

I prepared a PR for fixing that - https://github.com/algolia/algoliasearch-magento-2/pull/87
I'd really appreciate if you could test it on your enterprise editions. Thanks!

**Comment by JanPetr** [comment]
Fixed in [1.0.3](https://github.com/algolia/algoliasearch-magento-2/releases/tag/1.0.3)

**Comment by mhistihori** [comment]
Still got this error when reindex even use the latest version (1.0.3) with Magento 2.1 EE.

> Algolia Search Products indexer process unknown error:
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'backend.entity_id' in 'field list', query was: SELECT CONCAT(backend.store_id, '-', backend.entity_id), `backend`.`value` FROM `catalog_category_entity_varchar` AS `backend`
 INNER JOIN `catalog_category_entity` AS `category` ON backend.entity_id = category.entity_id AND (category.created_in <= 1 AND category.updated_in > 1) WHERE (backend.attribute_id = '45') AND (category.level > 1)
Algolia Search Categories indexer process unknown error:
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'backend.entity_id' in 'field list', query was: SELECT CONCAT(backend.store_id, '-', backend.entity_id) AS `key`, `category`.`path`, `backend`.`value` FROM `catalog_category_entity_int` AS `backend`
 INNER JOIN `catalog_category_entity` AS `category` ON backend.entity_id = category.entity_id AND (category.created_in <= 1 AND category.updated_in > 1) WHERE (backend.attribute_id = '46') ORDER BY `backend`.`store_id` ASC, `backend`.`entity_id` ASC

**Comment by JanPetr** [comment]
@mhistihori #111

**Comment by Xiris42** [comment]
@mhistihori @JanPetr

I didn't think to update this thread. 

The provided fix in 1.0.3 is not working on Magento EE 2.1+ because in the BaseHelper.php, getMagentoEdition() is called, which leads to : 

`return $this->productMetadata->getEdition();`

 and the getEdition(); is returning the Class Constant :

`const EDITION_NAME  = 'Community';`

The EDITION_NAME is the same for Magento CE and EE, and, correct me if I'm wrong, it's a preference that modifies it for the Enterprise Version.

You can fix easily, the same way we did, by creating a plugin afterGetEdition() where you can adjust it as you want. Feel free to check my collegue`s answer : https://github.com/algolia/algoliasearch-magento-2/pull/87#issuecomment-259479726

Xi-
